packages/gatsby/src/internal-plugins/component-layout-creator/gatsby-node.js

```diff
@@ -0,0 +1,70 @@
+const globCB = require(`glob`)
+const Promise = require(`bluebird`)
+const _ = require(`lodash`)
+const chokidar = require(`chokidar`)
+const systemPath = require(`path`)
+
+const glob = Promise.promisify(globCB)
+
+const createPath = require(`./create-path`)
+const validatePath = require(`./validate-path`)
+
+// Path creator.
+// Auto-create layouts.
+// algorithm is glob /layouts directory for js/jsx/cjsx files *not*
+// underscored. Then create url w/ our path algorithm *unless* user
+// takes control of that page component in gatsby-node.
+exports.createPagesStatefully = async (
+  { store, boundActionCreators },
+  options,
+  doneCb
+) => {
+  const { createLayout, deleteLayout } = boundActionCreators
+  const program = store.getState().program
+  const layoutDirectory = systemPath.posix.join(
+    program.directory,
+    `/src/layouts`
+  )
+  const exts = program.extensions.map(e => `${e.slice(1)}`).join(`,`)
+
+  // Get initial list of files.
+  let files = await glob(`${layoutDirectory}/**/?(${exts})`)
+  files.forEach(file => _createLayout(file, layoutDirectory, createLayout))
+
+  // Listen for new component pages to be added or removed.
+  chokidar
```

Comment:
Unrelated but I wonder if we should centralize file watching so we don't end up with lots of overlapping chokidar instances